### PR TITLE
chore: hand off pending push delivery on foreground, not at init

### PR DESCRIPTION
**Pull request by mahmoud-elmorabea**
[MBL-1753: Explore on-device deduping for Android identify calls](https://linear.app/customerio/issue/MBL-1753/explore-on-device-deduping-for-android-identify-calls)

## Goal

Make each push-delivered metric reach the Customer.io backend **exactly once**, choosing the channel by app foreground state:

- **Background-only process** (FCM-woken): WorkManager + direct HTTP is the only credible channel — use it.
- **User opens the app** (foreground): the analytics pipeline becomes available with reliable foreground network and full Segment storage. Anything WorkManager hasn't confirmed by then is handed off to the analytics pipeline, and WorkManager is told to stop trying.

Invariants:
1. **Channel exclusivity** — exactly one of WorkManager or the analytics pipeline delivers each event.
2. **No silent loss** — a local-persistence entry is removed only after one channel has confirmed delivery or been explicitly cancelled in favour of the other.

## What changed

- New `PendingDeliveryStore<T>` + `Serializer<T>` in `core/`, taking over the disk-backed JSON queue plumbing (`ReentrantLock`, oldest-by-timestamp eviction, corruption-safe reads, write-only-when-changed). `messagingpush` now feeds entries through it; the `location` module can adopt the same store in a follow-up.
- `ModuleMessagingPushFCM` no longer flushes orphans from `initialize()`. It registers a `ProcessLifecycleOwner` observer on the main thread; the handoff fires on every `ON_START` (re-fires across background → foreground cycles, with an empty-store fast-path).
- The handoff order per entry is `workManager.cancelUniqueWork(deliveryId).await()` → `eventBus.publish(...)` → final `removeAll(snapshottedKeys)`. Cancel happens before publish so a worker in `ENQUEUED`/`RUNNING` flips to `CANCELLED` before the analytics-pipeline event runs.
- L1–L10 INFO log moments added under tag `CIO-Push-Handoff` so behaviour is observable from `adb logcat`.
- Sample app gains a dev-only `SimulatePushDeliveryReceiver` for verification scenarios (drives the production `CustomerIOFirebaseMessagingService.onMessageReceived` path from a synthetic `RemoteMessage`).

## Out of scope (future work)

- **Geofence module (WS3)** — same pattern needs to land in `location/`, removing the parallel `eventBus.publish` from `GeofenceBroadcastReceiver`. Separate PR after this one.
- Direct-HTTP async fallback (used when WM is unavailable at enqueue time), foreground-time live-emission shortcutting, backend-side dedup — all unchanged, all out of scope.

## Verification

Ran on samples/kotlin_compose against a connected emulator. Log moments collected via `adb logcat -d | grep CIO-Push-Handoff`. Cold-start broadcasts use a dev-only intent that drives the production path. The emulator's failure to reach the real Customer.io host produced `UnknownHostException` retries naturally, exercising both branches.

| ID | Setup | Trigger | Observed sequence |
|---|---|---|---|
| P1-like | Force-stopped, network up | 1 broadcast (`refg-d1`) | append → enqueue WM → **worker success, removed**. No handoff fires for this key. |
| P2 | Force-stopped, host unreachable | 1 broadcast (`cold-d1`), wait, then activity launch | append → enqueue WM → worker retry. Then **on activity start**: foreground reached → cancelled WM → published to eventBus → handoff complete. |
| P3 | App already foreground | 1 broadcast (`p3-foreground`) | append → enqueue WM → worker success, removed. No handoff. |
| P4 | Force-stopped | 3 broadcasts (`p4-d1/d2/d3`), then activity launch | 3 × (append + enqueue). WM retried d1+d2, succeeded for d3 (removed). On foreground: snapshot count=**2** (d3 already gone), cancel + publish + remove for each. |
| Re-foreground | Foregrounded → home → cold push → foreground again | broadcast (`refg-d1`), home key, foreground | WM succeeded in background. On re-foreground: snapshot count=0 (empty-store fast-path). |

**Channel-exclusivity invariant held for every key**: `worker success, removed` count + `published to eventBus` count = 1.
**Store-file dump after each scenario**: `cio_pending_push_delivery.json` is `[]`.

Reproducing locally:
```
./gradlew :samples:kotlin_compose:installDebug
adb shell am force-stop io.customer.android.sample.kotlin_compose
adb logcat -c
adb shell am broadcast \
  -a io.customer.sample.SIMULATE_PUSH_DELIVERY \
  --es deliveryId d1 --es token tok-1 \
  -n io.customer.android.sample.kotlin_compose/.util.SimulatePushDeliveryReceiver \
  -f 0x20000000
# wait a few seconds, then:
adb shell am start -n io.customer.android.sample.kotlin_compose/.MainActivity
adb logcat -d | grep CIO-Push-Handoff
```
Note: requires the sample's debug-mode preference toggled on so SDK log level reaches DEBUG; otherwise the INFO lines are filtered.

## Test plan

- [x] `./gradlew :core:testDebugUnitTest` — new `PendingDeliveryStoreTest` (capacity, locking, corruption safety, etc.).
- [x] `./gradlew :messagingpush:testDebugUnitTest` — `ModuleMessagingPushFCMTest` covers init-does-not-flush, cancel-then-publish ordering, empty-store fast-path, and WM-unavailable async-tracker fallback. `PendingPushDeliveryMetricSerializerTest` covers JSON roundtrip + malformed-row skipping.
- [x] `./gradlew :datapipelines:testDebugUnitTest` — confirms downstream module unaffected.
- [x] `./gradlew apiCheck` — `@InternalCustomerIOApi` marker keeps the new class out of the public surface.
- [x] `./gradlew assembleDebug` — full project + sample app builds.
- [x] P1-like / P2 / P3 / P4 / re-foreground scenarios on the emulator.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Medium risk because it changes when/how push delivery metrics are flushed and adds WorkManager cancellation to enforce single-delivery; incorrect lifecycle or cancellation ordering could cause duplicate or dropped metric delivery.
> 
> **Overview**
> **Moves pending push-metric flushing from module init to app foreground.** `ModuleMessagingPushFCM` now observes `ProcessLifecycleOwner` and on each `ON_START` snapshots pending entries, cancels their unique WorkManager jobs, publishes `TrackPushMetricEvent` to the analytics pipeline, then removes only the snapshotted keys.
> 
> **Introduces a reusable disk-backed pending queue in `core`.** Adds generic `PendingDeliveryStore<T>` with serializer-driven JSON persistence, bounded oldest-first eviction, locking, and corruption-safe “no change, no write” semantics; `messagingpush` migrates from the deleted `PendingPushDeliveryStore` to `PendingDeliveryStore<PendingPushDeliveryMetric>` with a new JSON `Serializer`.
> 
> Adds structured INFO logging under `CIO-Push-Handoff`, updates workers/schedulers and tests to cover cancel-then-publish ordering and empty-store paths, and adds a sample-only `SimulatePushDeliveryReceiver` plus a new `androidx.lifecycle` process-lifecycle dependency for verification.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 82acad2d583b7d6324e1e18a8927036dbefc6d57. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: core/src/main/kotlin/io/customer/sdk/data/store/PendingDeliveryStore.kt [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+    private val serializer: Serializer<T>,
+    private val logger: Logger,
+    private val maxEntries: Int = DEFAULT_MAX_ENTRIES
+) {
+    /**
+     * Per-feature description of how to persist a [T] entry, what key to use
+     * for [remove]/[removeAll], and what timestamp drives oldest-first
+     * eviction.
+     */
+    interface Serializer<T> {
```
Comment (mahmoud-elmorabea) [inline core/src/main/kotlin/io/customer/sdk/data/store/PendingDeliveryStore.kt]: Do we really need our own `Serializer` interface here? Check rest of the sdk if there is something we can reuse or even Gson's type. If the only reason for this is the `key`, create an upper bound for `T` to a type that requires a string key

File: messagingpush/src/main/java/io/customer/messagingpush/ModuleMessagingPushFCM.kt [hunk 2]
```diff
@@ -60,32 +75,72 @@ class ModuleMessagingPushFCM @JvmOverloads constructor(
...
+                if (pending.isEmpty()) {
+                    pushLogger.logForegroundSnapshot(count = 0)
+                    return@runCatching
+                }
+                pushLogger.logForegroundSnapshot(count = pending.size)
 
-                val flushedIds = pending.map { it.deliveryId }
+                val wm = workManagerForCancel
                 pending.forEach { entry ->
+                    wm?.cancelUniqueWork(entry.deliveryId)?.await()
```
Comment (mahmoud-elmorabea) [inline messagingpush/src/main/java/io/customer/messagingpush/ModuleMessagingPushFCM.kt]: Can await throw exceptions? If so, it shouldn't short-circuit the remaining items

File: messagingpush/src/main/java/io/customer/messagingpush/ModuleMessagingPushFCM.kt [hunk 3]
```diff
@@ -60,32 +75,72 @@ class ModuleMessagingPushFCM @JvmOverloads constructor(
...
+                        handoffPendingPushDeliveryToAnalyticsPipeline()
+                    }
+                }
+            )
+        }
+        if (Looper.myLooper() == Looper.getMainLooper()) {
+            attach.run()
+        } else {
+            Handler(Looper.getMainLooper()).post(attach)
+        }
```
[the remark this answers is not in the document]
Comment (mahmoud-elmorabea) [reply messagingpush/src/main/java/io/customer/messagingpush/ModuleMessagingPushFCM.kt:111]: `initialize()` is only ever called once

File: messagingpush/src/main/java/io/customer/messagingpush/di/DiGraphMessagingPush.kt [hunk 4]
```diff
@@ -53,10 +54,12 @@ internal val SDKComponent.deepLinkUtil: DeepLinkUtil
 val SDKComponent.pushTrackingUtil: PushTrackingUtil
     get() = newInstance<PushTrackingUtil> { PushTrackingUtilImpl() }
 
-internal val SDKComponent.pendingPushDeliveryStore: PendingPushDeliveryStore
-    get() = singleton<PendingPushDeliveryStore> {
-        PendingPushDeliveryStore(
+internal val SDKComponent.pendingPushDeliveryStore: PendingDeliveryStore<PendingPushDeliveryMetric>
```
Comment (mrehan27) [inline messagingpush/src/main/java/io/customer/messagingpush/di/DiGraphMessagingPush.kt:57]: Should we create a dedicated class or typealias for `PendingDeliveryStore<PendingPushDeliveryMetric>` to make this easier to read and reuse?

File: messagingpush/src/main/java/io/customer/messagingpush/ModuleMessagingPushFCM.kt [hunk 5]
```diff
@@ -60,32 +75,94 @@ class ModuleMessagingPushFCM @JvmOverloads constructor(
...
+     * if initialize() is invoked off-main, we post to the main looper first.
+     */
+    private fun observeProcessForeground() {
+        val newObserver = LifecycleEventObserver { _: LifecycleOwner, event: Lifecycle.Event ->
+            if (event == Lifecycle.Event.ON_START) {
+                handoffPendingPushDeliveryToAnalyticsPipeline()
+            }
+        }
+        val attach = Runnable {
+            val processLifecycle = ProcessLifecycleOwner.get().lifecycle
```
Comment (mrehan27) [inline messagingpush/src/main/java/io/customer/messagingpush/ModuleMessagingPushFCM.kt:102]: Not for this PR, but `ProcessLifecycleOwner.get()` is now being used in a few classes. At some point, we should probably extract something common around it for consistency and easier handling.

File: messagingpush/src/main/java/io/customer/messagingpush/ModuleMessagingPushFCM.kt [hunk 6]
```diff
@@ -60,32 +75,94 @@ class ModuleMessagingPushFCM @JvmOverloads constructor(
...
+            foregroundObserver?.let { processLifecycle.removeObserver(it) }
+            foregroundObserver = newObserver
+            processLifecycle.addObserver(newObserver)
+        }
+        if (Looper.myLooper() == Looper.getMainLooper()) {
+            attach.run()
+        } else {
+            Handler(Looper.getMainLooper()).post(attach)
```
Comment (mrehan27) [inline messagingpush/src/main/java/io/customer/messagingpush/ModuleMessagingPushFCM.kt:110]: Should we use the existing `MainThreadPoster` here?

File: samples/java_layout/src/debug/java/io/customer/android/sample/java_layout/util/SimulatePushDeliveryReceiver.kt [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+ * path runs exactly as it would for a real FCM message.
+ *
+ * Usage:
+ *   adb shell am broadcast \
+ *     -a io.customer.sample.SIMULATE_PUSH_DELIVERY \
+ *     --es deliveryId d1 --es token tok1 \
+ *     -n io.customer.android.sample.java_layout/.util.SimulatePushDeliveryReceiver \
+ *     -f 0x20000000
+ */
+class SimulatePushDeliveryReceiver : BroadcastReceiver() {
```
Comment (mrehan27) [inline samples/java_layout/src/debug/java/io/customer/android/sample/java_layout/util/SimulatePushDeliveryReceiver.kt:24]: This is purely for local manual testing and doesn't trigger anywhere else in the sample app, right?

**Review by mrehan27 — APPROVED** [review]
Looks good 👍🏻

File: samples/java_layout/src/debug/java/io/customer/android/sample/java_layout/util/SimulatePushDeliveryReceiver.kt (hunk 7 above)
Comment (mahmoud-elmorabea) [reply samples/java_layout/src/debug/java/io/customer/android/sample/java_layout/util/SimulatePushDeliveryReceiver.kt:24]: Yes, this allows us to test some corner cases that are not possible with fly. Like duplicate messages delivery...etc

File: messagingpush/src/main/java/io/customer/messagingpush/di/DiGraphMessagingPush.kt (hunk 4 above)
Comment (mahmoud-elmorabea) [reply messagingpush/src/main/java/io/customer/messagingpush/di/DiGraphMessagingPush.kt:57]: I'd rather make it explicit so it's readable where this is coming from

File: messagingpush/src/main/java/io/customer/messagingpush/ModuleMessagingPushFCM.kt (hunk 5 above)
Comment (mahmoud-elmorabea) [reply messagingpush/src/main/java/io/customer/messagingpush/ModuleMessagingPushFCM.kt:102]: I will extract this along with main looper util in the next PR where we do extraction on geo-fencing branch
